### PR TITLE
Change `mylen` function to account for NaN empty lists when exploding on multiple columns

### LINKED ISSUE #46084: BUG: DataFrame.explode incomplete support on multiple columns with NaN or empty lists
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np

df = pd.DataFrame({"A": [[0, 1], [5], [], [2, 3]],"B": [9, 8, 7, 6],"C": [[1, 2], np.nan, [], [3, 4]],})

df.explode(["A", "C"])
```

### Issue Description

The `DataFrame.explode` function has incomplete behavior for the above example when exploding on multiple columns. For the above DataFrame `df = pd.DataFrame({"A": [[0, 1], [5], [], [2, 3]],"B": [9, 8, 7, 6],"C": [[1, 2], np.nan, [], [3, 4]],})` the outputs for exploding columns "A" and "C" individual are correctly outputted as
```python3
>>> df.explode("A")
     A  B       C
0    0  9  [1, 2]
0    1  9  [1, 2]
1    5  8     NaN
2  NaN  7      []
3    2  6  [3, 4]
3    3  6  [3, 4]

>>> df.explode("C")
        A  B    C
0  [0, 1]  9    1
0  [0, 1]  9    2
1     [5]  8  NaN
2      []  7  NaN
3  [2, 3]  6    3
3  [2, 3]  6    4
```
However, when attempting `df.explode(["A", "C"])`, one receives the error
```python3
>>> df.explode(["A", "C"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/emaanhariri/opt/anaconda3/lib/python3.9/site-packages/pandas/core/frame.py", line 8254, in explode
    raise ValueError("columns must have matching element counts")
ValueError: columns must have matching element counts
```
Here the `mylen` function on line 8335 in the [explode function](https://github.com/pandas-dev/pandas/blob/v1.4.1/pandas/core/frame.py#L8221-L8348) is defined as `mylen = lambda x: len(x) if is_list_like(x) else -1` induces the error above as entries `[]` and `np.nan` are computed as having different lengths despite both ultimately occupying a single entry in the expansion. 

### Expected Behavior

We expect the output to be an appropriate mix between the output of `df.explode("A")` and `df.explode("C")` as follows.

```python3 
>>> df.explode(["A", "C"])
     A  B    C
0    0  1    1
0    1  1    2
1    5  7  NaN
2  NaN  2  NaN
3    3  4    1
3    4  4    2
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.8.12.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 21.3.0
Version          : Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T6000
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.4
numpy            : 1.20.3
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 58.0.4
Cython           : 0.29.24
pytest           : 6.2.5
hypothesis       : None
sphinx           : 4.4.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.7.1
html5lib         : None
pymysql          : 1.0.2
psycopg2         : None
jinja2           : 3.0.2
IPython          : 7.29.0
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : None
fsspec           : 2021.11.1
fastparquet      : None
gcsfs            : 2021.11.1
matplotlib       : 3.4.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 5.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.7.3
sqlalchemy       : 1.4.27
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.55.0

</details>

### PR DESCRIPTION
**Pull request by ehariri**
- [x] closes #46084 
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by mroeschke** [comment]
Thanks for the pull request. Doesn't this fix relate to an existing issue? Can you demonstrate the behavior before/after fix?

File: pandas/tests/frame/methods/test_explode.py [hunk 1]
```diff
@@ -223,9 +223,15 @@ def test_explode_sets():
 
 
 @pytest.mark.parametrize(
-    "input_subset, expected_dict, expected_index",
+    "input_dict, input_index, input_subset, expected_dict, expected_index",
     [
         (
+            {
```
Comment (njriasan) [inline pandas/tests/frame/methods/test_explode.py:229]: I think you should include the github issue on both of your test changes to make sure its clear why you added these tests.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -8616,7 +8616,7 @@ def explode(
         if len(columns) == 1:
             result = df[columns[0]].explode()
         else:
-            mylen = lambda x: len(x) if is_list_like(x) else -1
+            mylen = lambda x: len(x) if is_list_like(x) and len(x) else 1
```
Comment (rhshadrach) [inline pandas/core/frame.py:8619]: nit: for readability, can you make `and len(x) > 0`.

File: pandas/tests/frame/methods/test_explode.py [hunk 3]
```diff
@@ -260,18 +272,32 @@ def test_explode_sets():
...
+            },
+            list("aaabcdeefg"),
+        ),
     ],
 )
-def test_multi_columns(input_subset, expected_dict, expected_index):
+def test_multi_columns(
+    input_dict, input_index, input_subset, expected_dict, expected_index
+):
     # GH 39240
```
Comment (rhshadrach) [inline pandas/tests/frame/methods/test_explode.py:299]: In regards to https://github.com/pandas-dev/pandas/pull/46791/files#r851832909, add the issue number here; i.e. `# GH 39240, 46084`

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: pandas/tests/frame/methods/test_explode.py [hunk 4]
```diff
@@ -223,9 +223,15 @@ def test_explode_sets():
 
 
 @pytest.mark.parametrize(
-    "input_subset, expected_dict, expected_index",
+    "input_dict, input_index, input_subset, expected_dict, expected_index",
```
Comment (phofl) [inline pandas/tests/frame/methods/test_explode.py:226]: Could you make new tests? This is really hard to review

**Review by phofl — COMMENTED** [review]
Could you please add a whatsnew note?

**Review by jreback — CHANGES_REQUESTED** [review]
great, can you add a whatsnew note. bug fixes for 1.5, in the reshaping section.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**